File: C++ Implementations/src/linear_structures/csll.cpp
#include "linear_structures/csll.h"

// ==== Helper Function ====
void CSLL::free_nodes() {
    if (!head) return;
    SharedNode* cursor = head;
    do {
        SharedNode* next_node = cursor->get_next();
        cursor->set_next(nullptr);
        delete cursor;
        cursor = next_node;
    } while (cursor && cursor != head);
}

// ==== Legacy ====

// Constructor
CSLL::CSLL() = default;

// Desctructor
CSLL::~CSLL() {
    // Deleting all the DAM nodes
    CSLL::free_nodes();

    // Resetting the pointers
    head = nullptr;
    tail = nullptr;
    no_of_elements = 0;
}

// ==== Private Non - Member Function ====
// Overloaded extraction operator to manage the display of the list.
std::ostream& operator<<(std::ostream& os, const CSLL& csll) {
    if (csll.no_of_elements != 0) {
        auto cursor = csll.head;
        while(cursor) {
            if (cursor == csll.head) {
                os << cursor->get_value() << " <= Head" << std::endl;
            }
            else if (cursor == csll.tail) {
                os << cursor->get_value() << " <= Tail" << std::endl;
                break;
            }
            else {
                os << cursor->get_value() << std::endl;
            }
            cursor = cursor->get_next();
        }
    } 
    else {
        os << "None <= Head" << std::endl;
        os << "None <= Tail" << std::endl;
    }

    return os;
}

void CSLL::append(int value) {
    // Adds a new node to the end of the list.
    auto new_node = new SharedNode(value);

    // If there are no elements present in the list.
    if (no_of_elements == 0) {
        head = new_node;
        tail = new_node;
        new_node->set_next(new_node);
    }
    // If there are elements present in the list.
    else {
        tail->set_next(new_node);
        tail = new_node;
        new_node->set_next(head);
    }
    // Updating the no of elements in the list.
    no_of_elements += 1;
}

void CSLL::insert(int value, int index) {
    // Adds a new node at a given index in the list.
    // Bounds Check
    if ((index < 0) || (index > no_of_elements)) {
        throw std::out_of_range("The requested index is out of range.");
    }
    else {
        auto new_node = new SharedNode(value);

        // Inserting element at the start of the list
        if (index == 0) {
            new_node->set_next(head);
            head = new_node;
            tail->set_next(new_node);
        }
        // Inserting the element at the end of the list
        else if (index == no_of_elements) {
            CSLL::append(value);
        }
        // Inserting the element at any random location
        else {
            auto cursor = head;
            for (int i = 0; i < index - 1; i++) {
                cursor = cursor->get_next();
            }
            new_node->set_next(cursor->get_next());
            cursor->set_next(new_node);
        }
        // Updating the no of elements in the list.
        no_of_elements += 1;
    }
}

int CSLL::search(int value) {
    // Searches for a given value in the list and returns the index of the first find.
    if (!head) {
        return -1;
    }
    auto cursor = head;
    int i = 0;
    while(true) {
        // If hit then returns the index
        if (cursor->get_value() == value) {
            return i;
        }
        if (cursor == tail) {
            break;
        }
        // Updates the loop control variables
        cursor = cursor->get_next();
        i += 1;
    }
    // If loop completes then no hit returns -1
    return -1;
}

SharedNode* CSLL::get(int index) {
    // Returns a reference to the node requested
    // Bounds Check
    if ((index < 0) || (index >= no_of_elements)) {
        throw std::out_of_range("The requested index is out of range.");
    }
    // Traversing to the requested index
    auto cursor = head;
    for (int i = 0; i < index; i++) {
        cursor = cursor->get_next();
    }
    return cursor;
}

void CSLL::set(int value, int index) {
    // Updates the value of a given node identified by its index.
    auto update_node = CSLL::get(index);
    update_node->set_value(value);
}

SharedNode* CSLL::pop() {
    // Removes and returns the last node of the list.
    if (!head) {
        throw std::out_of_range("The list is empty no more nodes can be popped.");
    }
    SharedNode* popped_node = nullptr;
    if (no_of_elements == 1) {
        popped_node = head;
        head = nullptr;
        tail = nullptr;
    }
    else {
        auto previous_ele = CSLL::get(no_of_elements - 2);
        popped_node = tail;
        previous_ele->set_next(head);
        tail = previous_ele;
    }
    // Update the no of elements
    no_of_elements -= 1;
    popped_node->set_next(nullptr);
    return popped_node;
}

void CSLL::remove(int index) {
    // Removes the node identified by its index.
    if ((index < 0) || (index >= no_of_elements)) {
        throw std::out_of_range("The requested index is out of bounds.");
    }
    else {
        SharedNode* remove_node = nullptr;

        // If the node to be removed is the first node.
        if (index == 0) {
            remove_node = head;
            SharedNode* next_node = head->get_next();
            tail->set_next(next_node);
            head = next_node;
        }
        // If the node to be removed is the last node.
        else if (index == no_of_elements - 1) {
            CSLL::pop();
            return;
        }
        // Removing any random node in the middle.
        else {
            remove_node = CSLL::get(index);
            SharedNode* previous_node = CSLL::get(index - 1);
            previous_node->set_next(remove_node->get_next());
        }

        // Updating the links of the remove node
        remove_node->set_next(nullptr);

        // Updating the no of elements in the list.
        no_of_elements -= 1;
    }
}

void CSLL::clear() {
    // Clears all the elements in the list.
    CSLL::free_nodes();
    head = nullptr;
    tail = nullptr;
    no_of_elements = 0;
}

File: Python Implementations/linear_structures/csll.py
from typing import Any, Generator


# Will add generics after the complete impl.
class Node:
    """Class that implements the Nodes that compose the elements of the CSLL."""
    # ==== Standard Methods ====
    def __init__(self, value: Any):
        self.value = value
        self.next: Node | None = None

    def __str__(self):
        return f"Node(value={self.value})"


class CSLL:
    """Class that implements all the behaviour of the Circular Singly-Linked List."""
    
    # ==== Standard Methods ====
    def __init__(self) -> None:
        
        self.head: Node | None = None
        self.tail: Node | None = None
        self.no_of_elements: int = 0

    def __str__(self) -> str:
        """Provides a string repr of the list of nodes."""
        
        list_str = ""
        # If the list is empty
        if not self.head:
            list_str += "None <= Head\nNone <= Tail\n"
        # If there is only one element in the list
        elif self.no_of_elements == 1:
            list_str += f"{self.head} <= Head\n{self.tail} <= Tail\n"
        # If the list is not empty
        else:
            cursor = self._head
            while cursor:
                if cursor == self.head:
                    list_str += f"{cursor.value} <= Head\n"
                elif cursor == self.tail:
                    list_str += f"{cursor.value} <= Tail\n"
                    break
                else:
                    list_str += f"{cursor.value}\n"
                # Updating the loop control variable for traversal
                cursor = cursor.next

        return list_str
    
    def __iter__(self) -> Generator[Node, Any, Any]:
        """Provides an iterator for the CSLL."""
        
        if not self._head:
            raise IndexError("The list is current empty.")
        else:
            cursor: Node = self._head
            while cursor:
                yield cursor

                # Exit Condition for Infinite Loop
                if cursor == self._tail:
                    break

                # Update the loop control variable
                assert cursor.next is not None, "Cursor cannot be none during traversal."
                cursor = cursor.next

    # ==== Helper Properties to update the pointers ====
    @property
    def _head(self) -> Node:
        
        assert self.head is not None, "The head is currently none."
        return self.head
    
    @property
    def _tail(self) -> Node:
        
        assert self.tail is not None, "The tail is currently none."
        return self.tail
    
    # ==== Properties of the CSLL ====
    def append(self, value: int) -> None:
        """Adds a new node to the end of the list."""
        
        new_node: Node = Node(value)
        
        # If the list is empty
        if not self.head:
            self.head = new_node
            self.tail = new_node
            new_node.next = self.head
        # If the list is not empty
        else:
            tail = self._tail
            tail.next = new_node
            new_node.next = self.head
            self.tail = new_node

        # Updating the no of elements
        self.no_of_elements += 1

    def insert(self, value: int, index: int) -> None:
        """Inserts the value at a given index in the list."""
        
        if index < 0 or index >= self.no_of_elements:
            raise IndexError("The requested index is out of bounds.")
        else:
            # Adds the node at the start of the list.
            new_node: Node = Node(value)
            if index == 0:
                new_node.next = self._head
                self.head = new_node
                self._tail.next = self.head
            # Adds the element at the end of the list.
            elif index == self.no_of_elements - 1:
                self.append(value)
                return
            # Adds the element at any random index of the list.
            else:
                cursor = self._head
                # Traversing to the idx - 1 to add the new node.
                for _ in range(index - 1):
                    assert cursor.next is not None, "Cursor cannot be None during traversal."
                    cursor = cursor.next
                
                # Updating the links
                assert cursor.next is not None, "The next ptr cannot be None in the middle of the list."
                new_node.next = cursor.next
                cursor.next = new_node

            # Updating the no of elements
            self.no_of_elements += 1

    def search(self, value: int) -> int:
        """Returns a integer after searching for a given value in the list."""
        
        for idx, node in enumerate(self):
            if node.value == value:
                return idx
        return -1
    
    def get(self, index: int) -> Node | None:
        """Returns the node at a given index of the list."""
        
        if index < 0 or index >= self.no_of_elements:
            raise IndexError("The requested index is out of bounds for the given list.")
        else:
            for idx, node in enumerate(self):
                if idx == index:
                    return node
            else:
                return None
    
    def set(self, value: int, index: int) -> None:
        """Updates the value of given index of the list."""
        
        if index < 0 or index >= self.no_of_elements:
            raise IndexError("The requested index is out of bounds for the given list.")
        else:
            # Accessing the corresponding Node 
            update_node = self.get(index)

            # If no Node was found
            assert isinstance(update_node, Node), "No node found at given index."

            # Update Node
            update_node.value = value
    
    def pop(self) -> Node:
        """Removes the very last node from the list and returns it."""
        
        if not self.head:
            raise IndexError("The list is currently empty and has no elements that can be popped.")
        else:
            # If only one element is left in the list
            remove_node: Node | None = None
            if self.no_of_elements == 1:
                remove_node = self._head
                self.head = None
                self.tail = None
            else:
                remove_node = self._tail
                previous_node = self.get(self.no_of_elements - 2)
                assert previous_node is not None, "The previous node cannot be none in the middle of the list."
                previous_node.next = self._head
                remove_node.next = None
                self.tail = previous_node
        
            # Updating the no of elements in the list.
            self.no_of_elements -= 1
            return remove_node
    
    def remove(self, index: int) -> None:
        """Removes any node identified by its index in the list."""
        
        if index < 0 or index >= self.no_of_elements:
            raise IndexError("The requested index was out of bounds.")
        else:
            remove_node: Node | None = None
            if index == 0:
                remove_node = self._head
                self.head = self._head.next
                self._tail.next = self.head
                remove_node.next = None
            elif index == self.no_of_elements - 1:
                self.pop()
                return
            else:
                remove_node = self.get(index)
                assert remove_node is not None, "The remove node cannot be none within the bounds of the list."
                previous_node = self.get(index - 1)
                assert previous_node is not None, "The previous node cannot be none in the middle of the list."
                previous_node.next = remove_node.next
                remove_node.next = None

            # Updating the number of elements in the list
            self.no_of_elements -= 1
            del remove_node

            # Updating the pointers if the elements is 0
            if self.no_of_elements == 0:
                self.head = None
                self.tail = None
    
    def clear(self) -> None:
        """Clears all the nodes in the list."""
        
        self.head = None
        self.tail = None
        self.no_of_elements = 0
